includes/class-newspack-ads-gam.php

```diff
@@ -373,6 +664,239 @@ private static function serialize_ad_unit( $gam_ad_unit ) {
+	}
+
+	/**
+	 * Create a GAM Order.
+	 *
+	 * @param string $name          Order Name.
+	 * @param string $advertiser_id Order Advertiser ID.
+	 *
+	 * @return object Created order.
+	 */
+	public static function create_order( $name, $advertiser_id ) {
+		$order = new Order();
+		$order->setName( $name );
+		$order->setAdvertiserId( $advertiser_id );
+		$order->setTraffickerId( self::get_current_user()->getId() );
+		$service        = self::get_order_service();
+		$created_orders = $service->createOrders( [ $order ] );
+		return self::get_serialised_orders( $created_orders )[0];
+	}
+
+	/**
+	 * Create a GAM Creative.
+	 *
+	 * @param array[] $creatives_config Array of creative configurations.
+	 *
+	 * @return object Created creative.
+	 *
+	 * @throws \Exception If unable to create creatives.
+	 */
+	public static function create_creatives( $creatives_config = [] ) {
+		$creatives = [];
+		$xsi_types = [
+			'BaseDynamicAllocationCreative',
+			'BaseRichMediaStudioCreative',
+			'ClickTrackingCreative',
+			'HasDestinationUrlCreative',
+			'Html5Creative',
+			'InternalRedirectCreative',
+			'LegacyDfpCreative',
+			'ProgrammaticCreative',
+			'TemplateCreative',
+			'ThirdPartyCreative',
+			'UnsupportedCreative',
+			'VastRedirectCreative',
+		];
+		foreach ( $creatives_config as $creative_config ) {
+			$creative_config = wp_parse_args(
+				$creative_config,
+				[
+					'xsi_type' => 'Creative',
+				]
+			);
+			if ( ! in_array( $creative_config['xsi_type'], $xsi_types, true ) ) {
+				throw new \Exception( 'Invalid xsi type' );
+			}
+			$fully_qualified_creative_class = 'Google\\AdsApi\\AdManager\\' . self::GAM_API_VERSION . '\\' . $creative_config['xsi_type'];
+			$creative                       = new $fully_qualified_creative_class();
+			$creative->setName( $creative_config['name'] );
+			$creative->setAdvertiserId( $creative_config['advertiser_id'] );
+			$creative->setSize( new Size( $creative_config['width'], $creative_config['height'] ) );
+			switch ( $creative_config['xsi_type'] ) {
+				case 'ThirdPartyCreative':
+					$creative_config = wp_parse_args(
+						$creative_config,
+						[
+							'snippet'                  => '',
+							'is_safe_frame_compatible' => true,
+						]
+					);
+					$creative->setSnippet( $creative_config['snippet'] );
+					$creative->setIsSafeFrameCompatible( $creative_config['is_safe_frame_compatible'] );
+					break;
+			}
+			$creatives[] = $creative;
+		}
+		$service           = self::get_creative_service();
+		$created_creatives = $service->createCreatives( $creatives );
+		return self::get_serialised_creatives( $created_creatives );
+	}
+
+	/**
+	 * Create line items.
+	 *
+	 * @param array[] $line_item_configs List of line item configurations.
+	 *
+	 * @return LineItem[] Created line items.
+	 *
+	 * @throws \Exception If unsupported configuration or unable to create line items.
+	 */
+	public static function create_line_items( $line_item_configs = [] ) {
```

Comment:
This is probably the trickiest part of the setup. Since we're unable to delete a created order we must be able to recover failed setups. It's hard to predict every scenario and we should improve at each new discovery.

The issue with your order is that the request failed in the middle of an association batch. I assumed that GAM wouldn't allow for a failed request to store the incomplete payload. To overcome that, d9b3df7fbee490c25d35e6754e58afa9b37c663b walks through the errors and unsets the duplicate associations from the batch to try again. Also added a message for the [`CONCURRENT_MODIFICATION` "common" error](https://developers.google.com/adwords/api/docs/common-errors#DatabaseError).

Reply:
I'm still able to get stuck in a limbo state pretty easily even with these changes. Perhaps instead of trying to delete any failed orders (which I now understand is impossible), we could "reset" things by allowing the user to enter a new order name and creating new line items, creatives, and LICAs under that order? And we could try archiving the failed order and possibly prefixing the name with something that makes it obvious that it shouldn't be used. WDYT?

Reply:
You're right! We can't predict all the possible API errors and should have a way out of the limbo. Here's what 38f291450818f4b6b88a905dd9d7b47fd669ab9f proposes:

 - Automatically set the order name to increment from existing, archived or not, orders on our Newspack Header Bidding advertiser
   - `Newspack Header Bidding` if it's the first order
   - `Newspack Header Bidding #2` for following orders
 - Allow for one fix attempt, which continues the step flow from where it failed
 - In case of a second failure, optimistically archive the order on GAM and removes it from the local database

![image](https://user-images.githubusercontent.com/820752/151574618-127277e0-d58f-4f27-af34-7d70eedbbc42.png)

You can reproduce this flow by adding a `throw new Exception();` to line `857` of `class-newspack-ads-bidding-gam.php`, which will cause every creative association attempt to fail.

@kmwilkerson, is this a good approach to handle unrecoverable order creations on GAM?

Reply:
Yes this is a great approach!! An alternative to #number would be version (v2, v3, etc). Referring to versions is common in ad ops 😃

Reply:
Cool! Updated in 0ca63ccab15cd9f686344e2f696813a90725eb04

Reply:
This recovery flow is really close to what I had in mind. It worked for me, with one caveat:

- I created the first order and canceled it during the line item creation step.
- I tried to "Fix Issues" and it failed again as describe above. The new error message gave me clear direction and the next action to take right away, with the incremented order name.
- I clicked "Create Order" again. This time I got a brief error message (didn't see any JS console errors or obvious error responses from API requests in the Network tab):

<img width="709" alt="Screen Shot 2022-01-28 at 10 50 34 AM" src="https://user-images.githubusercontent.com/2230142/151597166-4e8e230d-d90c-48db-a83e-e09eb3a6d617.png">

- Clicking "Create Order" again started the process over as expected, and this time I was able to let it finish.

The "not a valid JSON response" error was slightly concerning, but I'd consider it non-blocking since it still allowed me to try again.